packages/terminal/src/browser/terminal-linkmatcher-files.ts

```diff
@@ -0,0 +1,172 @@
+/********************************************************************************
+ * Copyright (C) 2019 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { ApplicationServer } from '@theia/core/lib/common/application-protocol';
+import { OS } from '@theia/core/lib/common';
+import { OpenerService } from '@theia/core/lib/browser';
+import { Position } from '@theia/editor/lib/browser';
+import { AbstractCmdClickTerminalContribution } from './terminal-linkmatcher';
+import { TerminalWidgetImpl } from './terminal-widget-impl';
+import { Path } from '@theia/core';
+import { FileSystem } from '@theia/filesystem/lib/common';
+import URI from '@theia/core/lib/common/uri';
+
+@injectable()
+export class TerminalLinkmatcherFiles extends AbstractCmdClickTerminalContribution {
+
+    protected backendOs: Promise<OS.Type>;
+
+    constructor(
+        @inject(ApplicationServer) appServer: ApplicationServer,
+        @inject(OpenerService) protected openerService: OpenerService,
+        @inject(FileSystem) protected fileSystem: FileSystem,
+    ) {
+        super();
+        this.backendOs = appServer.getBackendOS();
+    }
+
+    async getRegExp(): Promise<RegExp> {
+        const os = await this.backendOs;
+        const baseLocalLinkClause = os === OS.Type.Windows ? winLocalLinkClause : unixLocalLinkClause;
+        return new RegExp(`${baseLocalLinkClause}(${lineAndColumnClause})`);
+    }
+
+    getValidate(terminalWidget: TerminalWidgetImpl): (link: string) => Promise<boolean> {
+        return async match => {
+            const toOpen = await this.toURI(match, await terminalWidget.cwd);
+            if (toOpen) {
+                // TODO: would be better to ask the opener service, but it returns positively even for unknown files.
+                const f = await this.fileSystem.getFileStat(toOpen.toString());
+                return f !== undefined && !f.isDirectory;
```

Comment:
Strange, this is `null` on Windows.

```
terminal-linkmatcher-files.ts:54 Uncaught (in promise) TypeError: Cannot read property 'isDirectory' of null
    at TerminalLinkmatcherFiles.<anonymous> (terminal-linkmatcher-files.ts:54)
    at step (terminal-linkmatcher-files.ts:15)
    at Object.next (terminal-linkmatcher-files.ts:15)
    at fulfilled (terminal-linkmatcher-files.ts:15)
(anonymous) @ terminal-linkmatcher-files.ts:54
step @ terminal-linkmatcher-files.ts:15
(anonymous) @ terminal-linkmatcher-files.ts:15
fulfilled @ terminal-linkmatcher-files.ts:15
Promise.then (async)
step @ terminal-linkmatcher.ts:15
(anonymous) @ terminal-linkmatcher.ts:15
push.../../packages/terminal/lib/browser/terminal-linkmatcher.js.__awaiter @ terminal-linkmatcher.ts:15
validationCallback @ terminal-linkmatcher.ts:53
_loop_1 @ Linkifier.ts:246
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._doLinkifyRow @ Linkifier.ts:207
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._linkifyRows @ Linkifier.ts:122
(anonymous) @ Linkifier.ts:87
setTimeout (async)
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier.linkifyRows @ Linkifier.ts:87
push.../../node_modules/xterm/lib/Terminal.js.Terminal._queueLinkification @ Terminal.ts:1215
(anonymous) @ Terminal.ts:688
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
(anonymous) @ Terminal.ts:838
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._renderRows @ Renderer.ts:206
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer._innerRefresh @ RenderDebouncer.ts:56
(anonymous) @ RenderDebouncer.ts:42
requestAnimationFrame (async)
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer.refresh @ RenderDebouncer.ts:42
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.refreshRows @ Renderer.ts:197
push.../../node_modules/xterm/lib/Terminal.js.Terminal.refresh @ Terminal.ts:1204
push.../../node_modules/xterm/lib/renderer/CursorRenderLayer.js.CursorRenderLayer.onFocus @ CursorRenderLayer.ts:79
(anonymous) @ Renderer.ts:158
(anonymous) @ Renderer.ts:182
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._runOperation @ Renderer.ts:182
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.onFocus @ Renderer.ts:158
(anonymous) @ Terminal.ts:776
push.../../node_modules/xterm/lib/common/EventEmitter.js.EventEmitter.emit @ EventEmitter.ts:73
push.../../node_modules/xterm/lib/Terminal.js.Terminal._onTextAreaFocus @ Terminal.ts:582
(anonymous) @ Terminal.ts:752
terminal-linkmatcher-files.ts:54 Uncaught (in promise) TypeError: Cannot read property 'isDirectory' of null
    at TerminalLinkmatcherFiles.<anonymous> (terminal-linkmatcher-files.ts:54)
    at step (terminal-linkmatcher-files.ts:15)
    at Object.next (terminal-linkmatcher-files.ts:15)
    at fulfilled (terminal-linkmatcher-files.ts:15)
(anonymous) @ terminal-linkmatcher-files.ts:54
step @ terminal-linkmatcher-files.ts:15
(anonymous) @ terminal-linkmatcher-files.ts:15
fulfilled @ terminal-linkmatcher-files.ts:15
Promise.then (async)
step @ terminal-linkmatcher.ts:15
(anonymous) @ terminal-linkmatcher.ts:15
push.../../packages/terminal/lib/browser/terminal-linkmatcher.js.__awaiter @ terminal-linkmatcher.ts:15
validationCallback @ terminal-linkmatcher.ts:53
_loop_1 @ Linkifier.ts:246
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._doLinkifyRow @ Linkifier.ts:207
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier._linkifyRows @ Linkifier.ts:122
(anonymous) @ Linkifier.ts:87
setTimeout (async)
push.../../node_modules/xterm/lib/Linkifier.js.Linkifier.linkifyRows @ Linkifier.ts:87
push.../../node_modules/xterm/lib/Terminal.js.Terminal._queueLinkification @ Terminal.ts:1215
(anonymous) @ Terminal.ts:688
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
(anonymous) @ Terminal.ts:838
push.../../node_modules/xterm/lib/common/EventEmitter2.js.EventEmitter2.fire @ EventEmitter2.ts:46
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer._renderRows @ Renderer.ts:206
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer._innerRefresh @ RenderDebouncer.ts:56
(anonymous) @ RenderDebouncer.ts:42
requestAnimationFrame (async)
push.../../node_modules/xterm/lib/ui/RenderDebouncer.js.RenderDebouncer.refresh @ RenderDebouncer.ts:42
push.../../node_modules/xterm/lib/renderer/Renderer.js.Renderer.refreshRows @ Renderer.ts:197
push.../../node_modules/xterm/lib/Terminal.js.Terminal.refresh @ Terminal.ts:1204
push.../../node_modules/xterm/lib/Terminal.js.Terminal._onTextAreaBlur @ Terminal.ts:600
(anonymous) @ Terminal.ts:753

```

Reply:
So, I have checked why it fails on Windows; I do not know why does the `getFileStat` resolve to `null` for an invalid URI on Windows but the URI is incorrect. This is the `toOpen.toString()`.
```
file:///c%3A/Users/kittaakos/Desktop/xxx/c%3A%5CUsers%5Ckittaakos%5CDesktop%5Cxxx
```
My workspace root on Windows is:
```
C:\Users\kittaakos\Desktop\xxx
```

I hope this helps.

Reply:
btw, it's the same on docker/linux.

Reply:
Filed https://github.com/theia-ide/theia/issues/5200